waku/v2/node/peer_manager/waku_peer_store.nim

```diff
@@ -62,12 +67,41 @@ type
     connectedness*: Connectedness
     disconnectTime*: int64
     origin*: PeerOrigin
-    direction*: Direction
+    direction*: PeerDirection
+    lastFailedConn*: Moment
+    numberFailedConn*: int
 
 ##################
 # Peer Store API #
 ##################
 
+proc canBeConnected*(peerStore: PeerStore,
+                     peerId: PeerId,
+                     initialBackoffInSec: int,
+                     backoffFactor: int): bool =
+  # Returns if we can try to connect to this peer, based on past failed attempts
+  # It uses an exponential backoff. Each connection attempt makes us
+  # wait more before trying again.
+  let failedAttempts = peerStore[NumberFailedConnBook][peerId]
+
+  # if it never errored, we can try to connect
+  if failedAttempts == 0:
+    return true
+
+  # If it errored we wait an exponential backoff from last connection
+  # the more failed attemps, the greater the backoff since last attempt
+  let now = Moment.init(getTime().toUnix, Second)
+  let lastFailed = peerStore[LastFailedConnBook][peerId]
+  let backoff = chronos.seconds(initialBackoffInSec*(backoffFactor^(failedAttempts-1)))
+  if now >= (lastFailed + backoff):
+    return true
+  return false
+
+proc delete*(peerStore: PeerStore,
```

Comment:
Afaict, this is not yet used anywhere? Given the fact that in existing deployments peer IDs are cycled very often (I think), we should add a mechanism to manage the size of the peer store fairly urgently - unsure of the implication if this memory essentially leaks in the meantime.

Reply:
Yep, not used in this PR but tracked here in "Prune peers from peerstore".
shouldn't leak since its limited by `.withPeerStore(capacity=xxx)` but yes, we have to handle it more gracefully.